test/periphery/CardPaymentProcessorUpgradeable.test.ts

```diff
@@ -0,0 +1,1415 @@
+const BYTES32_LENGTH: number = 32;
+const ZERO_BYTES16_STRING: string = createBytesString('00', BYTES16_LENGTH);
+const ZERO_BYTES32_STRING: string = createBytesString('00', BYTES32_LENGTH);
+const SOME_BYTES16_STRING = createBytesString('ABCD', BYTES16_LENGTH);
+const SOME_BYTES32_STRING = createBytesString('DCBA', BYTES32_LENGTH);
+
+enum PaymentStatus {
+  Nonexistent = 0,
+  Uncleared = 1,
+  Cleared = 2,
+  Revoked = 3,
+  Reversed = 4,
+  Confirmed = 5,
+}
+
+interface TestPayment {
+  authorizationId: number,
+  account: SignerWithAddress,
+  amount: number,
+  status: PaymentStatus,
+  revocationCounter?: number,
+  makingPaymentCorrelationId: number
+  parentTxHash?: string
+}
+
+interface CardPaymentProcessorState {
+  tokenBalance: number;
+  totalClearedBalance: number,
+  totalUnclearedBalance: number,
+  clearedBalancesPerAccount: Map<string, number>,
+  unclearedBalancesPerAccount: Map<string, number>
+}
+
+function checkEquality(
+  actualOnChainPayment: any,
+  expectedPayment: TestPayment,
+  paymentIndex: number
+) {
+  if (expectedPayment.status == PaymentStatus.Nonexistent) {
+    expect(actualOnChainPayment.account).to.equal(
+      ethers.constants.AddressZero,
+      `payment[${paymentIndex}].account is incorrect`
+    );
+    expect(actualOnChainPayment.amount).to.equal(
+      0,
+      `payment[${paymentIndex}].amount is incorrect`
+    );
+    expect(actualOnChainPayment.amount).to.equal(
+      0,
+      `payment[${paymentIndex}].status is incorrect`
+    );
+    expect(actualOnChainPayment.revocationCounter).to.equal(0);
+  } else {
+    expect(actualOnChainPayment.account).to.equal(
+      expectedPayment.account.address,
+      `payment[${paymentIndex}].account is incorrect`
+    );
+    expect(actualOnChainPayment.amount).to.equal(
+      expectedPayment.amount,
+      `payment[${paymentIndex}].amount is incorrect`
+    );
+    expect(actualOnChainPayment.status).to.equal(
+      expectedPayment.status,
+      `payment[${paymentIndex}].status is incorrect`
+    );
+    expect(actualOnChainPayment.revocationCounter).to.equal(
+      expectedPayment.revocationCounter || 0,
+      `payment[${paymentIndex}].revocationCounter is incorrect`
+    );
+  }
+}
+
+describe("Contract 'CardPaymentProcessorUpgradeable'", async () => {
+  const REVERT_MESSAGE_IF_CONTRACT_IS_ALREADY_INITIALIZED = 'Initializable: contract is already initialized';
+  const REVERT_MESSAGE_IF_CALLER_IS_NOT_OWNER = "Ownable: caller is not the owner";
+  const REVERT_MESSAGE_IF_CALLER_IS_NEW_REVOCATION_COUNTER_MAXIMUM_VALUE_IS_ZERO =
+    "CardPaymentProcessor: new value of the revocation counter maximum must be greater than 0";
+  const REVERT_MESSAGE_IF_CONTRACT_IS_PAUSED = 'Pausable: paused';
+  const REVERT_MESSAGE_IF_PAYMENT_AMOUNT_IS_ZERO = 'CardPaymentProcessor: payment amount must be greater than 0';
+  const REVERT_MESSAGE_IF_INPUT_ARRAY_OF_AUTHORIZATION_IDS_IS_EMPTY =
+    'CardPaymentProcessor: input array of authorization IDs is empty';
```

Comment:
Codacy has a fix for the issue: [Replace `'CardPaymentProcessor:·input·array·of·authorization·IDs·is·empty'` with `"CardPaymentProcessor:·input·array·of·authorization·IDs·is·empty"`](https://app.codacy.com/gh/cloudwalk/brlc-token/pullRequest?prid=9781596)
```suggestion
    "CardPaymentProcessor: input array of authorization IDs is empty";
```